sdcm/remote/remote_libssh_cmd_runner.py

```diff
@@ -0,0 +1,138 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright (c) 2020 ScyllaDB
+
+from __future__ import absolute_import
+from typing import Optional, List
+import os
+import time
+import threading
+import socket
+
+from invoke.watchers import StreamWatcher
+
+from sdcm.utils.decorators import retrying
+
+from .libssh2_client import Client as LibSSH2Client, Timings
+from .libssh2_client.exceptions import AuthenticationException, UnknownHostException, ConnectError, \
+    FailedToReadCommandOutput, CommandTimedOut, FailedToRunCommand, OpenChannelTimeout, SocketRecvError, \
+    UnexpectedExit, Failure
+from .libssh2_client.result import Result
+from .base import RetryableNetworkException
+from .remote_base import RemoteCmdRunnerBase
+
+
+class RemoteLibSSH2CmdRunner(RemoteCmdRunnerBase, ssh_transport='libssh2'):  # pylint: disable=too-many-instance-attributes
+    """Remoter that mimic RemoteCmdRunner, under the hood it runs libssh2 client, instead of paramiko
+    Main problem in libssh2 - is that it is not thread safe, we mitigate this problem by having
+      _connection_thread_map - a dictionary in which we bind thread to the libssh2 session.
+    Whenever remoter read self.connection, we return value from _connection_thread_map associated with current thread,
+      And if it is not there, we create it.
+    """
+    connection_thread_map = threading.local()
+    exception_unexpected = UnexpectedExit
+    exception_failure = Failure
+    exception_retryable = (
+        # Exceptions that are not signaling about
+        AuthenticationException, UnknownHostException, ConnectError, FailedToReadCommandOutput,
+        CommandTimedOut, FailedToRunCommand, OpenChannelTimeout, SocketRecvError, socket.timeout
+    )
+
+    @property
+    def connection(self) -> LibSSH2Client:
+        """
+        Map connection to current thread.
+        If there is no such thread, create it.
+        """
+        connection = getattr(self.connection_thread_map, str(id(self)), None)
+        if connection is None:
+            connection = self._create_connection()
+            setattr(self.connection_thread_map, str(id(self)), connection)
+        return connection
+
+    @connection.setter
+    def connection(self, connection: LibSSH2Client):
+        setattr(self.connection_thread_map, str(id(self)), connection)
```

Comment:
Do you really need setter? It can be real headache to debug such thing because it's assigned to thread-local data.

Reply:
Not needed indeed. removed